[#9564] Consolidate Student Profile Picture RESTFul API

rrtheonlyone opened the pull request:
Part of #9564
Part of #9573
Part of #9595
Resolves #9574

**Changes**:

| New endpoint        | Method           | 
| ------------- |:-------------:| 
| /student/profilePic   | DELETE | 
|                     | PUT      |   
|                      | GET      | 
 
The biggest change here is with regards to the GET profile picture. Previously we used a `blobKey` to retrieve the picture. It has now been updated to use the student ID / email instead.

The authentication for the profile picture is as follows:
- **student** can view his own profile picture and his teammates' one
- **instructor** can view the profile picture of students in his course if he has sufficient privileges to do so

Tests have been done for the above.

Other changes as a result of this:
- slight improvement to the course details page (view photo previously used to remove the router params)

wkurniawan07 commented:
@rrtheonlyone You're going to find it difficult if you rely entirely on student email/course ID combination to get profile picture. Why not allow that in addition of still allowing usage of blob key?

rrtheonlyone commented:
@wkurniawan07 

The initial reason I went for this is because @xpdavid mentioned that the Blobkey is not encrypted and hence, may be a security risk if we use that for authentication. It also seemed cleaner to use the ID /email instead of passing the BlobKey back and forth. 

Do you suggest we allow both blobkey / id + email as a way to get the picture? (if blobkey is still used, how should we do the access control check?)

wkurniawan07 commented:
> Blobkey is not encrypted and hence, may be a security risk if we use that for authentication

The security concern is valid; which is why, unless authorized (e.g. getting self's profile), the blob key should not even be returned as API response. You'd realize this is the same case for regkeys.

> It also seemed cleaner to use the ID /email instead of passing the BlobKey back and forth.

I could tell you at least one place where there is no course ID and student email information. See if you can find it.

> Do you suggest we allow both blobkey / id + email as a way to get the picture? (if blobkey is still used, how should we do the access control check?)

Another similarity with regkey: if you have it, no further access control check is necessary, and rightly so.

rrtheonlyone commented:
Ready for review.

After some discussion, seems that the blobKey may not be needed (since there is sufficient information to retrieve picture). 

The authentication logic/flow needs to be looked into.

ChooJeremy commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -986,7 +986,7 @@ private Const() {
         public static final String COURSE_STUDENT_DETAILS = "/courses/students/details";
         public static final String STUDENT_COURSE = "/student/course";
-        public static final String STUDENT_PROFILE_PICTURE = "/students/profilePic";
+        public static final String STUDENT_PROFILE_PICTURE = "/student/profilePic";
```
Should probably standardize the name: `_` implies a `/`.
```suggestion
        public static final String STUDENT_PROFILEPIC = "/student/profilePic";
```

rrtheonlyone replied:
I really don't think this matters.

However, if the rest feel the same way I can change it :)

ChooJeremy commented on src/test/java/teammates/test/cases/webapi/GetStudentProfilePictureActionTest.java:
```diff
@@ -29,57 +26,195 @@ protected String getRequestMethod() {
-        GetStudentProfilePictureAction action = getAction(submissionParams);
-        assertThrows(NullHttpParameterException.class, () -> action.execute());
-    }
+        assertEquals(HttpStatus.SC_OK, imageResult.getStatusCode());
```
You should verify that the image result is correct instead of just the status code?
(Same scenarios for all of your other assertEquals)

rrtheonlyone replied:
Yup done!

junming403 commented on src/main/java/teammates/ui/webapi/action/GetStudentProfilePictureAction.java:
```diff
@@ -1,25 +1,73 @@
+            StudentAttributes student = logic.getStudentForEmail(courseId, studentEmail);
+
+            if (student == null) {
+                throw new UnauthorizedAccessException(UNAUTHORIZED_ACCESS);
```
This does not looks like `UnauthorizedAccessException` to me, should it be 404 instead?

rrtheonlyone replied:
Yes it is technically a 404. 

However, this is under the checkAccessControl method. This method checks access (it doesnt deal with action logic). By convention I have been throwing unauthorized access here.

rrtheonlyone replied:
Even in the `checkSpecificAccessControl` method?

That method is of `void` type and hence you cant return a `new JsonResult(..., HttpStatus.SC_NOT_FOUND)` as we normally do in our `Action`. As a result, for almost all of the actions here, I have returned unauthorized access if I cant find the student/instructor.

junming403 commented on src/main/java/teammates/ui/webapi/action/GetStudentProfilePictureAction.java:
```diff
@@ -1,25 +1,73 @@
+            studentProfile = logic.getStudentProfile(userInfo.id);
+        } else {
+            StudentAttributes student = logic.getStudentForEmail(courseId, studentEmail);
+            if (student == null) {
```
This has already been checked in access control parts?

rrtheonlyone replied:
Agreed! Removed

junming403 commented on src/main/java/teammates/ui/webapi/action/GetStudentProfilePictureAction.java:
```diff
@@ -1,25 +1,73 @@
+                throw new UnauthorizedAccessException(UNAUTHORIZED_ACCESS);
+            }
+
+            gateKeeper.verifyAccessibleForCurrentUserAsInstructorOrTeamMemberOrAdmin(logic.getAccount(userInfo.id),
```
Does admin has privilege on this api as well?

xpdavid replied:
We don't need to cater the case I think.

junming403 commented on src/test/java/teammates/test/cases/webapi/GetStudentProfilePictureActionTest.java:
```diff
@@ -29,57 +26,195 @@ protected String getRequestMethod() {
+        action = getAction(submissionParams);
+        imageResult = getImageResult(action);
+
+        assertEquals(HttpStatus.SC_OK, imageResult.getStatusCode());
```
Is there a way to verify the image is correct as well?

rrtheonlyone replied:
Yup done!

rrtheonlyone commented:
Ready for review.

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/GetStudentProfilePictureAction.java:
```diff
@@ -1,25 +1,70 @@
+        StudentProfileAttributes studentProfile = null;
+
+        if (studentEmail == null || courseId == null) {
+            if (!userInfo.isStudent) {
```
Yeap you are right! I have updated this.

rrtheonlyone commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.html:
```diff
@@ -132,7 +132,8 @@ <h4 class="font-weight-bold">Teammates</h4>
       <div class="row">
         <div class="col-md-2 col-3">
-          <img src="{{getPictureUrl(teammateProfile.pictureKey)}}" class="profile-pic float-left">
+          <img src="{{getPictureUrl(teammateProfile.pictureKey, teammateProfile.email)}}"
```
Yes you are right! We have to ensure this still works as intended.

Adoby7 commented on src/main/java/teammates/ui/webapi/action/GetStudentProfilePictureAction.java:
```diff
@@ -1,25 +1,71 @@
+                throw new UnauthorizedAccessException(UNAUTHORIZED_ACCESS);
+            }
+
+            gateKeeper.verifyAccessibleForCurrentUserAsInstructorOrTeamMemberOrAdmin(logic.getAccount(userInfo.id),
```
You can modify the `verifyAccessibleForCurrentUserAsInstructorOrTeamMemberOrAdmin` that takes a account id as a parameter. Since it only needs the id of account, there is no need to get the account attribute as we have already had the account id

rrtheonlyone replied:
Yeap, updated!

xpdavid commented on src/main/java/teammates/logic/api/GateKeeper.java:
```diff
@@ -396,19 +396,19 @@ private boolean isStudent() {
      * or his/her team member, or an admin.
      */
-    public void verifyAccessibleForCurrentUserAsInstructorOrTeamMemberOrAdmin(AccountAttributes account, String courseId,
+    public void verifyAccessibleForCurrentUserAsInstructorOrTeamMemberOrAdmin(String accountId, String courseId,
```
`accountId` -> googleId. This is a more suitable name.

rrtheonlyone replied:
Yes it is! I have updated it.

rrtheonlyone commented:
Ready for next stage of review.

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetStudentProfilePictureActionTest.java:
```diff
@@ -29,57 +27,170 @@ protected String getRequestMethod() {
+
+        verifyCanAccess(submissionParams);
+
+        ______TS("Success case: student can view his own photo but instructor or admin canno");
```
```suggestion
        ______TS("Success case: student can view his own photo but instructor or admin cannot");
```

rrtheonlyone replied:
Updated

xpdavid commented on src/web/app/pages-instructor/student-list/student-list.component.ts:
```diff
@@ -62,23 +57,17 @@ export class StudentListComponent implements OnInit {
+          student.photoUrl = (resp.status === 200)
+                ? `${this.backendUrl}/webapi/student/profilePic?courseid=${this.courseId}&studentemail=${student.email}`
+                : '/assets/images/profile_picture_default.png';
+        });
```
```suggestion
        }, handlerFor404);
```
should be put here?

rrtheonlyone replied:
Done

xpdavid commented on src/web/app/pages-instructor/student-profile/student-profile.component.html:
```diff
@@ -4,7 +4,7 @@
       <div class="row">
         <div class="col-md-2 col-3">
-          <img src="{{getPictureUrl(studentProfile?.pictureKey)}}" class="profile-pic float-right">
+          <img src="{{getPictureUrl(studentProfile.pictureKey)}}" class="profile-pic float-right">
```
You should use `(error) => ()` handler to update the picture to `/assets/images/profile_picture_default.png`?

The current logic is a bit weird and there will be no more pictureKey returning for studentProfile.

rrtheonlyone replied:
Yes you are right, I have made it better.

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.html:
```diff
@@ -132,7 +132,8 @@ <h4 class="font-weight-bold">Teammates</h4>
       <div class="row">
         <div class="col-md-2 col-3">
-          <img src="{{getPictureUrl(teammateProfile.pictureKey)}}" class="profile-pic float-left">
+          <img src="{{getPictureUrl(teammateProfile.pictureKey, teammateProfile.email)}}"
```
Same here.

rrtheonlyone replied:
Done!

rrtheonlyone commented:
@xpdavid I have updated the frontend logic to load default picture if no picture is found.

Ready for review.

rrtheonlyone commented:
Actually, I realised that the `student-profile` components still use the picture key (the box to upload/edit photos). Is it okay to leave it as is?

xpdavid commented on src/main/java/teammates/ui/webapi/action/GetStudentProfilePictureAction.java:
```diff
@@ -1,25 +1,71 @@
+            studentProfile = logic.getStudentProfile(userInfo.id);
+        } else {
+            StudentAttributes student = logic.getStudentForEmail(courseId, studentEmail);
+            studentProfile = logic.getStudentProfile(student.googleId);
```
The googleId can be null for legacy data. Use `String#isEmpty` to check?

rrtheonlyone replied:
Done

xpdavid commented on src/test/java/teammates/test/cases/webapi/GetStudentProfilePictureActionTest.java:
```diff
@@ -29,57 +27,170 @@ protected String getRequestMethod() {
+        assertEquals(HttpStatus.SC_OK, imageResult.getStatusCode());
+        assertEquals(student1Profile.pictureKey, imageResult.blobKey);
+
+        ______TS("Failure case: requested student has no profile picture");
```
Let's add a case where unregistered student's profile is requested?

rrtheonlyone replied:
Done

xpdavid commented on src/web/app/pages-instructor/student-profile/student-profile.component.ts:
```diff
@@ -12,26 +13,29 @@ import { StudentProfile } from './student-profile';
 export class StudentProfileComponent implements OnInit {
-
   Gender: typeof Gender = Gender; // enum
+  photoUrl?: string;
```
In this case, photoUrl can be an `@Input()` property?

An component should not have any assumption about the pages use them.

rrtheonlyone replied:
Yup! I have made it have a default value instead of being an optional. This isnt exactly an `@Input()` since it is always there.

xpdavid commented on src/web/app/pages-student/student-profile-page/student-profile-page.component.ts:
```diff
@@ -87,8 +87,8 @@ export class StudentProfilePageComponent implements OnInit {
     if (!pictureKey) {
```
pictureKey is still used here?

rrtheonlyone replied:
Updated

rrtheonlyone commented:
I have updated the API and also removed all use of `pictureKey` from the frontend.

Furthermore, I refactored the student profile component and profile picture upload modal. I removed unnecessary checks/methods and made the overall logic cleaner.

As a result, the user flow as a result has been improved (ref to #9573 and #9574). The process to upload a picture is now as follows:

![picture_upload_sm](https://user-images.githubusercontent.com/10579415/55292620-a75b2b00-541f-11e9-894a-ffa131ad2e08.gif)

It is now more intuitive to upload a picture. There are lesser buttons to click and the saving of profile picture is only done upon clicking of the upload button.

rrtheonlyone commented:
Unfortunately, it seems that this change in UX flow requires some change in the E2E tests as well (currently failing).

We can fix those if the new changes are approved.

xpdavid commented on src/web/app/pages-student/student-profile-page/upload-edit-profile-picture-modal/upload-edit-profile-picture-modal.component.ts:
```diff
@@ -18,146 +13,115 @@ import { StudentProfilePictureResults } from '../../../../types/api-output';
-  saveEditedPhoto(): void {
-    this.populateFormData(this.croppedImage);
-    this.uploadPicture();
+  populateFormData(file: File): void {
```
I am not sure why you need to call `populateFormData` every time when something changes. It looks like you can just construct the new formData in `uploadPicture()`?

It is confusing (I have to say, compared to the original) that `croppedImage ` and `formData` contain "same" data.

rrtheonlyone replied:
You have a good point here. 

However, there is a reason for this:
1. If an Image is changed in any way, I need access to the form data to send out the POST request. Populating it this way allows me to do it.
2. `croppedImage` has only one purpose. It preloads an image into the cropper if the user already has a profile picture. Other than that, I do not have any use for it. The library is built as such.

Do let me know if there is a better way to do this.

rrtheonlyone commented on src/main/java/teammates/ui/webapi/action/GetStudentProfilePictureAction.java:
```diff
@@ -1,25 +1,74 @@
+        } else {
+            StudentAttributes student = logic.getStudentForEmail(courseId, studentEmail);
+
+            if (!student.googleId.isEmpty()) {
```
I am not aware of `String.isEmpty()` method. Its what I have used above (from a non-static context) right? Or is there are a static method as well?

I have added an extra check here like this:
```
if (student.googleId != null && !student.googleId.isEmpty()) {
       studentProfile = logic.getStudentProfile(student.googleId);
}
```

rrtheonlyone commented on src/web/app/pages-instructor/student-profile/student-profile.component.ts:
```diff
@@ -12,26 +13,29 @@ import { StudentProfile } from './student-profile';
+  constructor(private route: ActivatedRoute) { }
 
   ngOnInit(): void {
+    this.route.queryParams.subscribe((queryParams: any) => {
```
Ahh sorry! You are right.

I have changed as per suggestion.

rrtheonlyone commented on src/web/app/pages-instructor/student-profile/student-profile.component.ts:
```diff
@@ -12,26 +13,29 @@ import { StudentProfile } from './student-profile';
 export class StudentProfileComponent implements OnInit {
-
   Gender: typeof Gender = Gender; // enum
+  photoUrl: string = '/assets/images/profile_picture_default.png';
```
Yup done!

rrtheonlyone commented on src/web/app/pages-student/student-course-details-page/__snapshots__/student-course-details-page.component.spec.ts.snap:
```diff
@@ -333,7 +333,7 @@ exports[`StudentCourseDetailsPageComponent should snap with populated fields 1`]
           >
             <img
               class="profile-pic float-left"
-              src="http://localhost:8080/webapi/students/profilePic?blob-key=something.jpg"
```
I have modified the logic slightly. The one where you commented on the `inline`.

As a result of this change, the image src is also different!

rrtheonlyone commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -100,19 +101,30 @@ export class StudentCourseDetailsPageComponent implements OnInit {
-      return '/assets/images/profile_picture_default.png';
-    }
-    return `${this.backendUrl}/webapi/students/profilePic?blob-key=${pictureKey}`;
+  loadPictureUrl(teammateProfile: TeammateProfile, courseId: string): void {
```
Yes your right! Updated.

rrtheonlyone commented on src/web/app/pages-student/student-profile-page/student-profile-page.component.ts:
```diff
@@ -56,8 +56,8 @@ export class StudentProfilePageComponent implements OnInit {
   editForm!: FormGroup;
   nationalities?: string[];
 
+  defaultPictureLink: string = '/assets/images/profile_picture_default.png';
```
I have actually used this in some other places also. I will modify all accordingly!

rrtheonlyone commented on src/web/app/pages-student/student-profile-page/student-profile-page.component.ts:
```diff
@@ -76,19 +76,17 @@ export class StudentProfilePageComponent implements OnInit {
-    }
-    this.currentTime = (new Date()).getTime(); // forces image reload in HTML template
-    return `${this.backendUrl}/webapi/students/profilePic?blob-key=${pictureKey}&time=${this.currentTime}`;
+  setDefaultPic(): void {
```
Done!

rrtheonlyone replied:
Done!

rrtheonlyone commented on src/web/app/pages-student/student-profile-page/upload-edit-profile-picture-modal/upload-edit-profile-picture-modal.component.ts:
```diff
@@ -18,146 +13,115 @@ import { StudentProfilePictureResults } from '../../../../types/api-output';
-        this.profilePicLink.replace(`${this.backendUrl}/webapi`, '');
-    return this.httpRequestService.get(profilePicEndPoint, {}, 'blob');
+  uploadPicture(): void {
+    this.activeModal.dismiss();
```
Yes good point. I thought about this but just went with the `Output` initially for convenience.

I have updated accordingly to make use of the method!

rrtheonlyone commented on src/web/app/pages-student/student-profile-page/upload-edit-profile-picture-modal/upload-edit-profile-picture-modal.component.ts:
```diff
@@ -18,146 +13,115 @@ import { StudentProfilePictureResults } from '../../../../types/api-output';
+  imageChangedEvent: any = '';
+  croppedImage: any = '';
+  showCropper: boolean = false;
+  imageLoadFailed: boolean = false;
```
Correct. This was an experimental flag I was playing around with. Apologies, will remove it since it feels like it is really not needed at the moment.

rrtheonlyone commented on src/web/app/pages-student/student-profile-page/upload-edit-profile-picture-modal/upload-edit-profile-picture-modal.component.ts:
```diff
@@ -18,146 +13,115 @@ import { StudentProfilePictureResults } from '../../../../types/api-output';
-  isFileSelected: boolean = false;
+  imageChangedEvent: any = '';
+  croppedImage: any = '';
+  showCropper: boolean = false;
```
Yup have removed this flag also.

rrtheonlyone commented on src/web/app/pages-student/student-profile-page/upload-edit-profile-picture-modal/upload-edit-profile-picture-modal.component.html:
```diff
@@ -6,70 +6,48 @@ <h4 class="modal-title">
+      [maintainAspectRatio]="false"
+      (imageCropped)="imageCropped($event)"
+      (imageLoaded)="imageLoaded()"
+      [style.display]="showCropper ? null : 'none'"
```
This part seems to do nothing as well. So, have removed it.

rrtheonlyone commented on src/web/app/pages-student/student-profile-page/upload-edit-profile-picture-modal/upload-edit-profile-picture-modal.component.ts:
```diff
@@ -18,146 +13,115 @@ import { StudentProfilePictureResults } from '../../../../types/api-output';
+  }
+
+  /**
+   * Checks if image is loaded
```
Removed method

rrtheonlyone commented:
Ready for review. 

I have improved the logic and also updated the tests.

xpdavid commented on src/web/app/pages-instructor/instructor-course-student-details-page/instructor-course-student-details-page.component.ts:
```diff
@@ -21,18 +22,25 @@ interface StudentDetails {
   studentProfile?: StudentProfile;
+  photoUrl: string = '';
+
+  private backendUrl: string = environment.backendUrl;
```
This can be a local variable?

rrtheonlyone replied:
Yup updated. I have just used `environment.backendUrl` directly in the profile picture link.

I have also changed all unnecessary uses of this variable in other classes.

xpdavid commented on src/web/app/pages-instructor/student-profile/student-profile.component.spec.ts:
```diff
@@ -9,6 +9,9 @@ describe('StudentProfileComponent', () => {
   beforeEach(async(() => {
     TestBed.configureTestingModule({
       declarations: [StudentProfileComponent],
+      imports: [
```
I guess you don't need this?

rrtheonlyone replied:
yes removed

xpdavid commented on src/web/app/pages-instructor/student-list/student-list.component.ts:
```diff
@@ -25,6 +20,8 @@ interface StudentDetails {
 })
 export class StudentListComponent implements OnInit {
 
+  readonly DEFAULT_PICTURE_LINK: string = '/assets/images/profile_picture_default.png';
```
Is `DEFAULT_PICTURE_LINK` used somewhere in the HTML template? I actually don't know which style do you want to follow:
1. export `DEFAULT_PICTURE_LINK ` to HTML template -> you don't need `setDefaultPic()` -> just use `(error)="student.photoUrl = DEFAULT_PICTURE_LINK;"`

2. Stick to `setDefaultPic()` -> delete export of `DEFAULT_PICTURE_LINK`

rrtheonlyone replied:
As of now, I dont have a fixed style. I have changed my code to follow (2) now. I can change it to (1) also if that is the convention. I personally prefer (2) as the method can be extended if needed (perhaps you want to do more actions on error for picture)

rrtheonlyone commented:
Made necessary changes - ready for review.

xpdavid commented on src/web/app/pages-instructor/student-profile/student-profile.component.spec.ts:
```diff
@@ -1,5 +1,5 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
-
+import { RouterTestingModule } from '@angular/router/testing';
```
I guess you can also remove this?

rrtheonlyone replied:
Yup

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -100,19 +98,23 @@ export class StudentCourseDetailsPageComponent implements OnInit {
+        const courseId: string = this.course.id;
+        this.teammateProfiles.forEach((teammateProfile: TeammateProfile) => teammateProfile.photoUrl
+          = `${environment.backendUrl}/webapi/student/profilePic?courseid=${courseId}
+          &studentemail=${teammateProfile.email}`);
```
The email is not the email in the teammteProfile. It is the email in `StudentAttributes`.

rrtheonlyone replied:
Ahh yess! Fixed.

xpdavid commented on src/web/app/pages-student/student-profile-page/student-profile-page.component.html:
```diff
@@ -8,7 +8,8 @@ <h1>{{ name }}</h1>
               class="profile-pic"
-              (click)="onUploadEdit()">
+              (click)="onUploadEdit()"
+              (error)="this.profilePicLink = this.DEFAULT_PICTURE_LINK">
```
`DEFAULT_PICTURE_LINK` is not defined?

rrtheonlyone replied:
Oh yah, updated! I will thoroughly check all my changes and make sure everything is working now :P

xpdavid commented on src/web/app/pages-student/student-profile-page/upload-edit-profile-picture-modal/upload-edit-profile-picture-modal.component.ts:
```diff
@@ -18,146 +13,106 @@ import { StudentProfilePictureResults } from '../../../../types/api-output';
+  imageChangedEvent: any = '';
   formData?: FormData;
-  imageToShow: any;
   croppedImage: any;
```
You can remove this?

rrtheonlyone replied:
Yup removed

xpdavid commented:
Please also check whether this PR resolve https://github.com/TEAMMATES/teammates/pull/9588 or not?

rrtheonlyone commented on src/e2e/java/teammates/e2e/cases/e2e/StudentProfilePageE2ETest.java:
```diff
@@ -51,15 +42,14 @@ public void testAll() {
-                BackDoor.getStudentProfile("SProfUiT.CS2104", "SProfUiT.student@gmail.tmt");
-        // checks that the pictureKey value is within the newly uploaded profile picture link
-        assertTrue(profilePage.getProfilePicLink().contains(studentProfileAttributes.pictureKey));
+        ______TS("Typical case: no picture");
```
Yup your right! Forgot to update the descriptions. Updated accordingly.

rrtheonlyone commented:
@xpdavid made the necessary changes!

Yes it should resolve #9588 

I have looked through the issue box and realised there are a number of issues which this PR should help fix (or make progress with). I will updated my main PR description accordingly to reference those.

xpdavid commented on src/web/app/pages-instructor/student-profile/student-profile.component.ts:
```diff
@@ -12,26 +11,20 @@ import { StudentProfile } from './student-profile';
 
   /**
-   * Construct the url for the profile picture from the given key.
+   * Sets the profile picture of a student as the default image
```
```suggestion
   * Sets the profile picture of a student as the default image.
```

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -100,19 +98,24 @@ export class StudentCourseDetailsPageComponent implements OnInit {
+      if (this.teammateProfiles && this.course && this.student) {
+        const courseId: string = this.course.id;
+        const studentEmail: string = this.student.email;
+        this.teammateProfiles.forEach((teammateProfile: TeammateProfile) => teammateProfile.photoUrl
```
The `photoUrl` is always the same?

I guess you should make use of the `/webapi/students` to get the student attributes of the team members.

rrtheonlyone replied:
Yup updated

xpdavid commented on src/web/app/pages-student/student-profile-page/upload-edit-profile-picture-modal/upload-edit-profile-picture-modal.component.html:
```diff
@@ -6,70 +6,46 @@ <h4 class="modal-title">
+  <div>
+    <image-cropper
+      [imageChangedEvent]="imageChangedEvent"
+      [imageBase64]="imageCropper.imageBase64"
```
Can this line be removed?

rrtheonlyone replied:
Yes it can! Removed

rrtheonlyone commented:
I have refactored the course details page to use the new APIs. 

The page now makes a few API calls namely:
1. `/course` : to get course details
2. `/student`: to get user details
3. `/students:`: to fetch teammates
4. `/student/profile`: to get student profiles
5. `/instructors`: to load instructors

A few things to note:
- The page UI/UX could use some work. The tables are not consistent and its not very visual
- The profile is entirely up to the user to fill in. Hence some of the tables have no content (user has blank profile). Might need to look into a better way of rendering this
- Lazy load teammates + profile in future? (upon button press)
- The teammates includes the current user as well.


Ready for next stage of review.

xpdavid commented on src/web/services/student.service.ts:
```diff
@@ -23,4 +23,15 @@ export class StudentService {
   }
+
+  /**
+   * Gets all students in a course and team as an instructor by calling API.
```
```suggestion
   * Gets all students in a course and team as a student by calling API.
```
right?

rrtheonlyone replied:
Yupp correct

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -1,46 +1,18 @@
-  course: CourseAttributes;
-  instructorDetails: InstructorDetails[];
-  teammateProfiles?: TeammateProfile[];
+interface StudentProfileWithPicture extends StudentProfile {
```
Prefer composition over inheritance. i.e. It is better to put student profile as a field inside the interface rather than extending it.

rrtheonlyone replied:
Yupp changed

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -52,67 +24,119 @@ interface StudentCourseDetails {
-  constructor(private route: ActivatedRoute, private httpRequestService: HttpRequestService,
+  student?: Student;
+  course?: Course;
+  instructorDetails?: Instructors;
```
Can just use `Instructor[]`?

rrtheonlyone replied:
Done

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -52,67 +24,119 @@ interface StudentCourseDetails {
-        this.statusMessageService.showErrorMessage('Error retrieving instructor details');
-      }
+  /**
+   * Loads the students of the course.
```
```suggestion
   * Loads the current logged-in student of the course.
```

rrtheonlyone replied:
Done

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -52,67 +24,119 @@ interface StudentCourseDetails {
+        .subscribe((students: Students) => {
+          students.students.forEach((student: Student) => {
+            // filter away current user
+            if (this.student && student.email === this.student.email && student.name === this.student.name) {
```
- Is this check necessary `this.student`?
- I think just checking email is enough.

rrtheonlyone replied:
Yup yourr right. Fixed

xpdavid commented on src/web/app/pages-student/student-course-details-page/__snapshots__/student-course-details-page.component.spec.ts.snap:
```diff
@@ -128,71 +133,6 @@ exports[`StudentCourseDetailsPageComponent should snap with populated fields 1`]
     </div>
   </div><div
     class="section"
-  >
```
I think you should modify the snapshot test accordingly?

rrtheonlyone replied:
Done!

rrtheonlyone commented:
Ready for review.

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -1,46 +1,22 @@
+/**
+ * A student profile which also has the profile picture URL
+ */
+export interface StudentProfileWithPicture {
```
You don't need to `export` it?

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.spec.ts:
```diff
@@ -36,48 +37,55 @@ describe('StudentCourseDetailsPageComponent', () => {
+      lastName: '1',
+      name: '1',
+      comments: '',
+      requestId: '5cb7d3320000b3b0cb4e5208',
```
You don't need to have construct requestId?

rrtheonlyone replied:
Yup removed.

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.spec.ts:
```diff
@@ -36,48 +37,55 @@ describe('StudentCourseDetailsPageComponent', () => {
+          moreInfo: 'Misc',
+          name: '2',
+          nationality: 'Andorran',
+          pictureKey: '',
```
You should remove `pictureKey` in `StudentProfileData`?

rrtheonlyone replied:
Removed and updated!

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.html:
```diff
@@ -111,60 +111,57 @@ <h4 class="font-weight-bold">Self</h4>
               <td class="font-weight-bold">Nationality</td>
               <td>
-                <span *ngIf="teammateProfile.nationality">{{teammateProfile.nationality}}</span>
+                <span *ngIf="teammateProfile.studentProfile.nationality">{{teammateProfile.studentProfile
```
I think you can safely remove the `*ngIf`. Check other places also.

rrtheonlyone replied:
Yup

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.html:
```diff
@@ -111,60 +111,57 @@ <h4 class="font-weight-bold">Self</h4>
 </div>
 
-<div class="section" *ngIf="teammateProfiles">
+<div class="section" *ngIf="teammateProfiles && teammateProfiles.length">
```
You don't need this check `teammateProfiles`?

I hope you know what the `teammateProfiles` in the condition is used for.

Same for other places.

rrtheonlyone replied:
Done! Yes I am aware. 

This was something I just reused from the current implementation. I have updated accordingly.

tanhengyeow commented on src/web/app/pages-instructor/instructor-course-student-details-page/instructor-course-student-details-page.component.ts:
```diff
@@ -21,18 +22,23 @@ interface StudentDetails {
   styleUrls: ['./instructor-course-student-details-page.component.scss'],
 })
 export class InstructorCourseStudentDetailsPageComponent implements OnInit {
-
```
Unrelated change here, most files do leave one line spacing before the declaration of variables so let's do that here as well :)

rrtheonlyone replied:
Done

tanhengyeow commented on src/web/app/pages-instructor/instructor-student-records-page/instructor-student-records-page.component.ts:
```diff
@@ -27,21 +28,26 @@ interface Session {
   styleUrls: ['./instructor-student-records-page.component.scss'],
 })
 export class InstructorStudentRecordsPageComponent implements OnInit {
-
```
Same here.

rrtheonlyone replied:
Done!

rrtheonlyone commented:
@tanhengyeow thanks for the review!

With regards to your comments:

**1/2**: Good catch! This is due to regression. I had two ways to update the picture `onError` initially and removed one! Now it is fixed.

Now for points 3-5, they seem very UX specific. There is a reason why I went to refactor out the profile picture component and redesign it. One, the logic was pretty complicated with heavy reliance on `pictureKey`. I had to remove this key as part of my migration and this meant a revamp of the page. Two, the UX was def not user friendly (refer to #9573) - it was difficult to upload a picture; you had to click like three buttons and the saving felt odd.

I just simplified the picture upload and made it obvious how to use! There is of course room for improvement. However, I think that is best left to a new issue and another PR. This PR has already: 
- touched on multiple pages to update Profile Picture API
- refactored modal component (to simplify upload process and improve flow) + updated E2E + snapshot
- redid API calls to course details (to get the profile picture URL) + updated snapshot

However, since you brought them up:

**3**: A user will open this modal and will be greeted with a smaller window (with some buttons and a message). The chance he wont read the message is pretty small right :P . Furthermore, why the need to complicate the process and make the buttons appear later? I do not see a great hindrance if the buttons are already there! However, this is just my opinion (up for discussion separately)

**4**: Yes, we can look into a better placement for this. (If I was as a user, I wont be too shocked f the message is close to the buttons; looks fine to me)

**5**: Yup sure! We can separately look into a separate flow for this. User only sees upload -> buttons + sentence

tanhengyeow commented:
Yup agree that the suggestions can be a new issue on its own as they are just some of my quick observations :smile: 

I can file one after this PR has been merged

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -52,67 +28,132 @@ interface StudentCourseDetails {
-      }
-    }, (resp: ErrorMessageOutput) => {
-      this.statusMessageService.showErrorMessage(resp.error.message);
+  loadCourse(courseid: string): void {
```
```suggestion
  loadCourse(courseId: string): void {
```

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -52,67 +28,132 @@ interface StudentCourseDetails {
+   * Loads the current logged-in student of the course.
+   * @param courseid: id of the course queried
+   */
+  loadStudent(courseid: string): void {
```
```suggestion
  loadStudent(courseId: string): void {
```

xpdavid commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.ts:
```diff
@@ -52,67 +28,132 @@ interface StudentCourseDetails {
+   * @param courseid: id of the course queried
+   * @param teamName: team of current student
+   */
+  loadTeammates(courseid: string, teamName: string): void {
```
```suggestion
  loadTeammates(courseId: string, teamName: string): void {
```

rrtheonlyone commented on src/web/app/pages-student/student-course-details-page/student-course-details-page.component.html:
```diff
@@ -36,15 +36,15 @@ <h4 class="font-weight-bold">Course</h4>
 </div>
 
-<div class="section" *ngIf="instructorDetails">
+<div class="section" *ngIf="instructorDetails && instructorDetails.length">
```
Actually, empty array is not regarded as false :O

Check this out: https://basarat.gitbooks.io/typescript/docs/javascript/truthy.html

wkurniawan07 replied:
Empty array (or empty object `{}`) is not `false`.

damithc commented on src/main/java/teammates/ui/webapi/action/GetStudentProfilePictureAction.java:
```diff
@@ -1,25 +1,75 @@
+    /** String indicating student not found. */
+    public static final String STUDENT_NOT_FOUND = "No student found for given parameters";
+
+    /** String indicating profile picture not found. */
```
Mentioning the type `String` in the comment is redundant?

rrtheonlyone replied:
Would `Message` be better?

Something like `/** Message indicating student not found. */`

damithc replied:
`Indicates the student is not found`?

rrtheonlyone replied:
Yup that is better! I have updated accordingly.

wkurniawan07 commented on src/web/app/pages-instructor/student-profile/student-profile.component.ts:
```diff
@@ -12,26 +11,20 @@ import { StudentProfile } from './student-profile';
-    }
-    return `${this.backendUrl}/webapi/students/profilePic?blob-key=${pictureKey}`;
+  setDefaultPic(): void {
+    this.photoUrl = '/assets/images/profile_picture_default.png';
```
This strengthens what I said about extracting this as constant.

rrtheonlyone replied:
Yes correct (used in a couple of places)

wkurniawan07 commented on src/e2e/java/teammates/e2e/cases/e2e/StudentProfilePageE2ETest.java:
```diff
@@ -51,15 +42,14 @@ public void testAll() {
-        logout();
+        profilePage.ensureProfileContains("short.name", "e@email.tmt", "inst", "American",
+                StudentProfileAttributes.Gender.FEMALE, "this is enough!$%&*</>");
+        profilePage.verifyStatusMessage(Const.StatusMessages.STUDENT_PROFILE_EDITED);
```
What are all the changes here?

rrtheonlyone replied:
These are not changes! Due to how selenium behaves, I had to reorder the test runner. 

The previous implementation was as follows:
1. Edit profile
2. Upload Picture
3. Edit picture
4. Ensure profile contains everything

However, with my redesign of this feature. This test repeatedly kept failing. The reason was due to the status bar verification. Once profile is edited, there will be a status bar message saying it is succesfully edited. After the profile picture is saved, due to Selenium's speed, the old status bar will still be around and it wont disappear for a couple of seconds. So this test will fail as instead of "Profile Picture Saved Succesfully", it is something like "Profile Picture edited succesfully.."

So my solution to fix (albeit hacky) this, was to reorder this process, I make it so that the profile picture is saved first, before the profile is edited. While the profile is being edited, there is enough time, for the status bar to disappear. Hence the changes.

rrtheonlyone commented on src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java:
```diff
@@ -155,9 +151,6 @@ public void editProfilePhoto() {
         click(editPictureFlipHorizontal);
-
-        click(editPictureSubmit);
-        waitForPageToLoad();
```
I dont have that button anymore! My "save" button will close off the modal.

rrtheonlyone commented on src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java:
```diff
@@ -177,12 +170,13 @@ public void showPictureEditor() {
-        assertEquals(width, browser.driver.findElement(By.className("cropper")).getCssValue("width"));
+        assertEquals(height, browser.driver.findElement(By.className("profile-pic")).getCssValue("height"));
+        assertEquals(width, browser.driver.findElement(By.className("profile-pic")).getCssValue("width"));
+        click(uploadEditModal.findElement(By.className("close")));
```
Yes you are right! This is wrong. 

I will fix this. I think this was something I was experimenting with earlier while debugging the E2E test.

rrtheonlyone commented on src/web/app/pages-instructor/student-list/student-list.component.ts:
```diff
@@ -62,23 +54,15 @@ export class StudentListComponent implements OnInit {
-    }
-    return `${this.backendUrl}/webapi/students/profilePic?blob-key=${pictureKey}`;
+  setDefaultPic(student: StudentListStudentData): void {
+    student.photoUrl = '/assets/images/profile_picture_default.png';
```
Yes exactly! Where though? I could not find any `const.ts` file to house constants like this.

amrut-prabhu replied:
How about [`config.ts`](https://github.com/TEAMMATES/teammates/blob/master/src/web/environments/config.template.ts)?

rrtheonlyone commented on src/web/app/pages-instructor/student-profile/student-profile.component.html:
```diff
@@ -4,7 +4,7 @@
       <div class="row">
         <div class="col-md-2 col-3">
-          <img src="{{getPictureUrl(studentProfile?.pictureKey)}}" class="profile-pic float-right">
+          <img src="{{this.photoUrl}}" class="profile-pic float-right" (error)="setDefaultPic()">
```
Nope I dont. I will update this as well.
